### ISSUE TITLE
Erroneous type extension

### ISSUE THREAD

**Issue by mk0x9** [microsoft/TypeScript#31236]
**TypeScript Version:**  3.5.0-dev.20190503

**Code**

```ts
type Foo = {
    prop?: string;
}

type FooFunction = (param: Foo) => Foo;

const lambda: FooFunction = (a) => {
    return { ...a, bar: 42 };
}

const c = lambda({});
```

**Expected behavior:** TypeScript compiler should throw an error here:
```
index.ts:8:20 - error TS2322: Type '{ bar: number; prop?: string; }' is not assignable to type 'Foo'.
  Object literal may only specify known properties, and 'bar' does not exist in type 'Foo'.

8     return { ...a, bar: 42 };
                     ~~~~~~~
```

**Actual behavior:** Compiles without any issue. In case if lambda defined as follows:

```ts
const lambda = (a: Foo): Foo => {
```

exception is thrown.

**Playground Link:** https://www.typescriptlang.org/play/#src=type%20Foo%20%3D%20%7B%0D%0A%20%20%20%20prop%3F%3A%20string%3B%0D%0A%7D%0D%0A%0D%0Atype%20FooFunction%20%3D%20(param%3A%20Foo)%20%3D%3E%20Foo%3B%0D%0A%0D%0Aconst%20lambda1%3A%20FooFunction%20%3D%20(a)%20%3D%3E%20%7B%0D%0A%20%20%20%20return%20%7B%20...a%2C%20bar%3A%2042%20%7D%3B%0D%0A%7D%0D%0A%0D%0Aconst%20lambda2%20%3D%20(a%3A%20Foo)%3A%20Foo%20%3D%3E%20%7B%0D%0A%20%20%20%20return%20%7B%20...a%2C%20bar%3A%2042%20%7D%3B%0D%0A%7D%0D%0A%0D%0Aconst%20c1%20%3D%20lambda1(%7B%7D)%3B%0D%0Aconst%20c2%20%3D%20lambda2(%7B%7D)%3B%0D%0A%0D%0A%2F%2F%20c1.bar%3F

**Comment by mk0x9** [comment]
Oops, dupe of #31235
